File: julia/heat1d.jl
# 1D Heat equation: u_t = alpha*u_xx
# Domain: [0,1], BC: u=0, IC: sin(pi*x)
# Exact: exp(-alpha*pi^2*t)*sin(pi*x)
# Method: explicit FTCS

using LinearAlgebra, Printf
using Plots; gr()

# ── parameters ────────────────────────────────────────────────────────────────
const alpha = 1.0
const N     = 50
const T     = 0.1
x  = LinRange(0, 1, N+2)
dx = x[2] - x[1]
r  = 0.4
dt = r * dx^2 / alpha
steps = Int(floor(T / dt)) + 1

# ── initial condition ─────────────────────────────────────────────────────────
u     = sin.(π .* x)
u[1]  = u[end] = 0.0
u_new = copy(u)

# ── snapshots ─────────────────────────────────────────────────────────────────
snap_steps = Set([1, div(steps,2), steps])
snaps = Dict{Float64, Vector{Float64}}()

# ── time march ────────────────────────────────────────────────────────────────
for n in 1:steps
    n in snap_steps && (snaps[n*dt] = copy(u))
    @inbounds for i in 2:N+1
        u_new[i] = u[i] + r*(u[i+1] - 2u[i] + u[i-1])
    end
    u_new[1] = u_new[end] = 0.0
    u, u_new = u_new, u
end

# ── exact & error ─────────────────────────────────────────────────────────────
exact(t) = exp(-alpha * π^2 * t) .* sin.(π .* x)
t_final  = (steps-1) * dt
l2_err   = sqrt(dx * sum((u .- exact(t_final)).^2))
@printf("1D Heat  N=%d  T=%.2f  L2 error = %.3e\n", N, T, l2_err)

# ── plot ──────────────────────────────────────────────────────────────────────
mkpath("../results")
p1 = plot(title="1D Heat (solid=num, dashed=exact)", xlabel="x", ylabel="u",
          legend=:topright)
colors = [:blue, :red, :green]
t_keys = sort(collect(keys(snaps)))
for (k, t_val) in enumerate(t_keys)
    plot!(p1, x, snaps[t_val], lw=2, color=colors[k], label="num t=$(round(t_val,digits=3))")
    plot!(p1, x, exact(t_val), lw=1, ls=:dash, color=colors[k], label=false)
end

p2 = plot(x, u .- exact(t_final), lw=1.5, color=:red,
          title=@sprintf("Error t=%.3f  L2=%.2e", t_final, l2_err),
          xlabel="x", ylabel="error", legend=false)

plt = plot(p1, p2, layout=(1,2), size=(1000,400),
           plot_title="Julia — 1D Heat Equation (FTCS)  N=$N")
savefig(plt, "../results/julia_heat1d.png")
println("Saved ../results/julia_heat1d.png")

File: matlab/heat1d.m
% 1D Heat equation: u_t = alpha*u_xx
% Domain: [0,1], BC: u(0)=u(1)=0, IC: sin(pi*x)
% Exact: exp(-alpha*pi^2*t)*sin(pi*x)
% Method: explicit FTCS

clear; close all;

alpha = 1.0;
N     = 50;
T     = 0.1;
x     = linspace(0, 1, N+2)';
dx    = x(2) - x(1);
r     = 0.4;
dt    = r * dx^2 / alpha;
steps = floor(T / dt) + 1;

u = sin(pi * x);
u(1) = 0; u(end) = 0;

snap_steps = [1, floor(steps/2), steps];
snaps = containers.Map('KeyType','double','ValueType','any');

for n = 1:steps
    if ismember(n, snap_steps)
        snaps(n * dt) = u;
    end
    u_new = u;
    u_new(2:end-1) = u(2:end-1) + r*(u(3:end) - 2*u(2:end-1) + u(1:end-2));
    u_new(1) = 0; u_new(end) = 0;
    u = u_new;
end

exact = @(t) exp(-alpha*pi^2*t) .* sin(pi*x);
t_final = (steps-1)*dt;
l2_err = sqrt(dx * sum((u - exact(t_final)).^2));
fprintf('1D Heat  N=%d  T=%.2f  L2 error = %.3e\n', N, T, l2_err);

% ── plot ──────────────────────────────────────────────────────────────────────
if ~exist('../results', 'dir'), mkdir('../results'); end

figure('Position', [100 100 1000 400]);
subplot(1,2,1); hold on;
colors = lines(length(snap_steps));
t_keys = sort(cell2mat(keys(snaps)));
for k = 1:length(t_keys)
    t_val = t_keys(k);
    plot(x, snaps(t_val), '-',  'Color', colors(k,:), 'LineWidth', 2, ...
         'DisplayName', sprintf('num t=%.3f', t_val));
    plot(x, exact(t_val), '--', 'Color', colors(k,:), 'LineWidth', 1);
end
xlabel('x'); ylabel('u'); title('1D Heat (solid=num, dashed=exact)');
legend('show', 'Location', 'best', 'FontSize', 7); grid on;

subplot(1,2,2);
plot(x, u - exact(t_final), 'r-', 'LineWidth', 1.5);
xlabel('x'); ylabel('error');
title(sprintf('Error at t=%.3f  L2=%.2e', t_final, l2_err)); grid on;

sgtitle(sprintf('MATLAB — 1D Heat Equation (FTCS)  N=%d', N));
saveas(gcf, '../results/matlab_heat1d.png');
fprintf('Saved ../results/matlab_heat1d.png\n');
